core/src/main/java/feign/querymap/FieldQueryMapEncoder.java

```diff
@@ -31,16 +37,20 @@
 public class FieldQueryMapEncoder implements QueryMapEncoder {
 
   private final Map<Class<?>, ObjectParamMetadata> classToMetadata =
-      new HashMap<Class<?>, ObjectParamMetadata>();
+      new ConcurrentHashMap<>();
 
   @Override
   public Map<String, Object> encode(Object object) throws EncodeException {
     try {
-      ObjectParamMetadata metadata = getMetadata(object.getClass());
-      Map<String, Object> fieldNameToValue = new HashMap<String, Object>();
+      AtomicReference<Object> atomicReference = new AtomicReference<>();
+      atomicReference.set(object);
+
+      ObjectParamMetadata metadata = getMetadata(atomicReference.get().getClass());
+
+      Map<String, Object> fieldNameToValue = new ConcurrentHashMap<>();
       for (Field field : metadata.objectFields) {
-        Object value = field.get(object);
-        if (value != null && value != object) {
+        Object value = field.get(atomicReference.get());
+        if (value != null && value != atomicReference.get()) {
```

Comment:
Instead of an `AtomicReference` you can use the concurrent hash maps `putIfAbsent`.  This will provide the atomic get/then/put we need without the reference.